File: app/runner.py
"""A/B Runner(设计文档 §4 / §10 / §12)。

S1:每 Provider 每题 1 次(同 prompt、同参数,仅切换 endpoint/model);
S2:只对 A/B 结果不一致的分歧题,每端再跑 repeat_disagreements 次,
    形成 3-run majority(文档 §12 的"只复测 disagreement"成本控制)。
"""
from __future__ import annotations

import asyncio
import traceback
from datetime import datetime, timezone
from typing import Optional

from . import analyzer
from .evaluator import evaluate_heuristic
from .provider import LiveProvider, build_prompt
from .schemas import Instance, ProviderConfig
from .store import RunStore, records_to_csv

_INSTANCE_FIELDS = set(Instance.model_fields.keys())

# run_id → 原始 api_key(仅进程内,不落盘)
SECRET_VAULT: dict[str, dict[str, str]] = {}


def _to_instance(d: dict) -> Instance:
    return Instance(**{k: v for k, v in d.items() if k in _INSTANCE_FIELDS})


def compute_cost(pconf: ProviderConfig, prompt_tokens: int,
                 cached_tokens: int, completion_tokens: int) -> float:
    uncached = max(0, prompt_tokens - cached_tokens)
    return (uncached / 1e6 * pconf.price_input_per_m
            + cached_tokens / 1e6 * pconf.price_cached_per_m
            + completion_tokens / 1e6 * pconf.price_output_per_m)


class RunEngine:
    def __init__(self, store: RunStore):
        self.store = store

    async def execute(self, run_id: str) -> None:
        state = self.store.load_state(run_id)
        if state is None:
            return
        request_data = state["request"]
        # api_key 已脱敏存储,真实 key 通过进程内缓存回填
        req = _restore_request(request_data, SECRET_VAULT.get(run_id, {}))

        suite = state["suite"]
        instances = suite["instances"]
        # 基线与候选端可独立运行:复用基线时基线记录直接导入,本次只实跑候选端
        providers: dict[str, ProviderConfig] = {}
        if req.baseline_run_id:
            providers["candidate"] = req.provider_b
        else:
            providers["baseline"] = req.provider_a
            if req.provider_b is not None:
                providers["candidate"] = req.provider_b

        try:
            self.store.update_state(run_id, status="running", phase="S1 主评测")
            if req.baseline_run_id:
                imported = self.store.import_baseline(run_id, req.baseline_run_id)
                st = self.store.load_state(run_id) or {}
                for rec in imported:
                    istat = st["instance_status"].setdefault(rec["instance_id"], {})
                    if rec["status"] != "ok":
                        badge = "error"
                    else:
                        badge = "pass" if rec["resolved"] else "fail"
                    istat[f"baseline_r{rec['run_index']}"] = badge
                    if rec["run_index"] == 0:
                        istat["baseline"] = badge
                self.store.update_state(run_id, instance_status=st["instance_status"])
            for inst in instances:
                for role, pconf in providers.items():
                    await asyncio.sleep(0)  # 让出事件循环,保证取消/查询可响应
                    await self._run_once(run_id, inst, pconf, role,
                                         run_index=0, phase="S1")
                    st = self.store.load_state(run_id) or {}
                    st["counts"]["s1"] = st["counts"].get("s1", 0) + 1
                    st["progress"]["done"] = st["counts"]["s1"] + st["counts"].get("s2", 0)
                    st["progress"]["current"] = None
                    self.store.update_state(run_id, **{
                        "counts": st["counts"], "progress": st["progress"]})

            # A/B 对比可用(双端同跑,或复用基线 + 实跑候选端)即触发分歧检测;
            # 复用基线时 providers 仅含 candidate,S2 只补跑候选端,基线保持导入结果
            if "candidate" in providers:
                records = self.store.load_records(run_id)
                disagreements = analyzer.disagreement_ids(records)
                if disagreements and req.repeat_disagreements > 0:
                    self.store.update_state(run_id, phase="S2 分歧复测", status="retesting")
                    by_id = {i["instance_id"]: i for i in instances}
                    st = self.store.load_state(run_id) or {}
                    st["progress"]["total"] = st["progress"]["done"] + \
                        len(disagreements) * len(providers) * req.repeat_disagreements
                    self.store.update_state(run_id, progress=st["progress"])
                    for iid in disagreements:
                        for run_index in range(1, req.repeat_disagreements + 1):
                            for role, pconf in providers.items():
                                await asyncio.sleep(0)
                                await self._run_once(run_id, by_id[iid], pconf, role,
                                                     run_index=run_index, phase="S2")
                                st = self.store.load_state(run_id) or {}
                                st["counts"]["s2"] = st["counts"].get("s2", 0) + 1
                                st["progress"]["done"] = (st["counts"]["s1"]
                                                          + st["counts"]["s2"])
                                self.store.update_state(run_id, counts=st["counts"],
                                                        progress=st["progress"])

            self.store.update_state(run_id, status="analyzing", phase="分析汇总")
            records = self.store.load_records(run_id)
            report = analyzer.build_report(run_id, records, suite, req)
            csv_text = records_to_csv(records)
            md_text = analyzer.render_markdown(report)
            self.store.write_report(run_id, report, csv_text, md_text)
            st = self.store.load_state(run_id) or {}
            st["progress"]["current"] = None
            self.store.update_state(run_id, status="completed", phase="完成",
                                    report_ready=True, progress=st["progress"],
                                    ended_at=datetime.now(timezone.utc)
                                    .strftime("%Y-%m-%dT%H:%M:%SZ"))
        except asyncio.CancelledError:
            self.store.update_state(run_id, status="cancelled", phase="已取消",
                                    ended_at=datetime.now(timezone.utc)
                                    .strftime("%Y-%m-%dT%H:%M:%SZ"))
            raise
        except Exception:
            self.store.update_state(run_id, status="failed", phase="失败",
                                    error=traceback.format_exc()[-2000:],
                                    ended_at=datetime.now(timezone.utc)
                                    .strftime("%Y-%m-%dT%H:%M:%SZ"))

    async def _run_once(self, run_id: str, inst_dict: dict, pconf: ProviderConfig,
                        role: str, run_index: int, phase: str) -> dict:
        inst = _to_instance(inst_dict)
        provider = LiveProvider(pconf)
        system, user = build_prompt(inst)
        started = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"

        st = self.store.load_state(run_id) or {}
        st["progress"]["current"] = {"instance_id": inst.instance_id, "role": role,
                                     "phase": phase}
        istat = st["instance_status"].setdefault(inst.instance_id, {})
        istat[f"{role}_r{run_index}"] = "running"
        self.store.update_state(run_id, progress=st["progress"],
                                instance_status=st["instance_status"])

        completion = await provider.complete(system, user)
        outcome = evaluate_heuristic(inst, completion)

        cost = compute_cost(pconf, completion.prompt_tokens,
                            completion.cached_tokens, completion.completion_tokens)
        record = {
            "run_index": run_index,
            "phase": phase,
            "instance_id": inst.instance_id,
            "provider_role": role,
            "provider_name": pconf.name,
            "model": pconf.model,
            "status": "ok" if completion.ok else "error",
            "resolved": (None if not completion.ok else bool(outcome.resolved)),
            "f2p_passed": outcome.f2p_passed, "f2p_total": outcome.f2p_total,
            "p2p_passed": outcome.p2p_passed, "p2p_total": outcome.p2p_total,
            "usage": {"prompt_tokens": completion.prompt_tokens,
                      "completion_tokens": completion.completion_tokens,
                      "cached_tokens": completion.cached_tokens},
            "ttft_s": completion.ttft_s,
            "wall_s": round(completion.wall_s, 3),
            "decode_tps": completion.decode_tps,
            "finish_reason": completion.finish_reason,
            "tool_errors": 0 if completion.ok else len(completion.errors),
            "errors": completion.errors,
            "cost_usd": round(cost, 6),
            "eval_method": outcome.method,
            "patch_excerpt": (completion.text or "")[:600],
            "started_at": started,
            "ended_at": datetime.now(timezone.utc)
                        .strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z",
        }
        self.store.append_record(run_id, record)

        st = self.store.load_state(run_id) or {}
        istat = st["instance_status"].setdefault(inst.instance_id, {})
        if completion.ok:
            badge = "pass" if outcome.resolved else "fail"
        else:
            badge = "error"
        istat[f"{role}_r{run_index}"] = badge
        istat[f"{role}"] = badge if run_index == 0 else istat.get(f"{role}", badge)
        self.store.update_state(run_id, instance_status=st["instance_status"])

        # 文档 §14:仅保存失败与分歧任务的完整轨迹,减少存储
        if badge in ("fail", "error"):
            self.store.save_trajectory(run_id, inst.instance_id, role, run_index, {
                "instance_id": inst.instance_id,
                "provider_role": role, "model": pconf.model,
                "run_index": run_index, "phase": phase,
                "eval_detail": outcome.detail,
                "prompt_system": system, "prompt_user": user,
                "response": completion.text, "record": record,
            })
        return record


def _restore_request(data: dict, secrets: dict) -> "RunRequest":
    """用进程内密钥缓存回填脱敏后的 api_key。"""
    from .schemas import ProviderConfig, RunRequest  # noqa: PLC0415

    def fix(p: Optional[dict], role: str) -> Optional[ProviderConfig]:
        if p is None:
            return None
        p = dict(p)
        key = secrets.get(role)
        if key:
            p["api_key"] = key
        return ProviderConfig(**p)

    return RunRequest(
        provider_a=fix(data["provider_a"], "baseline"),
        provider_b=fix(data.get("provider_b"), "candidate"),
        baseline_run_id=data.get("baseline_run_id"),
        suite_level=data.get("suite_level", "smoke6"),
        suite_seed=data.get("suite_seed", 0),
        repeat_disagreements=data.get("repeat_disagreements", 2),
        turn_limit=data.get("turn_limit", 50),
    )

File: src/runner.ts
import path from 'node:path';
import { RunStore, recordsToCsv } from './store';
import * as analyzer from './analyzer';
import * as livemod from './live';
import { AgentWorkspace, AGENT_TOOLS, AgentWorkspaceOptions } from './workspace';
import { evaluateOfficialDocker } from './evaluator';
import {
  LiveProvider,
  CompletionResult,
  createCompletionResult,
  buildPrompt,
} from './provider';
import {
  Instance,
  ProviderConfig,
  RunRequest,
  toInstance,
} from './schemas';

export const SECRET_VAULT = new Map<string, Record<string, string>>();

const AGENT_SYSTEM_TMPL =
  'You are an autonomous software engineer agent solving a real task inside a ' +
  'dedicated workspace directory.\n' +
  '- Interact with the workspace ONLY through the provided tools ' +
  '(bash / view_file / edit_file / submit).\n' +
  '- Create or modify files so that the final workspace state resolves the task. ' +
  'The submission patch is computed automatically from your file changes ' +
  '(git diff) — do NOT print the diff yourself.\n' +
  '- You have a budget of at most {turns} turns. Call `submit` when the task is done.\n' +
  '- Be efficient: inspect first, implement, then verify; avoid repeating ' +
  'failed commands.';

const BAND_ORDER: Record<string, number> = { easy: 0, medium: 1, hard: 2 };

function byDifficulty(instances: Array<Record<string, any>>): Array<Record<string, any>> {
  return [...instances].sort(
    (a, b) => (BAND_ORDER[a.difficulty] ?? 3) - (BAND_ORDER[b.difficulty] ?? 3),
  );
}

export function computeCost(
  pconf: ProviderConfig,
  promptTokens: number,
  cachedTokens: number,
  completionTokens: number,
): number {
  const uncached = Math.max(0, promptTokens - cachedTokens);
  return (
    (uncached / 1e6) * pconf.price_input_per_m +
    (cachedTokens / 1e6) * pconf.price_cached_per_m +
    (completionTokens / 1e6) * pconf.price_output_per_m
  );
}

export class RunEngine {
  constructor(private readonly store: RunStore) {}

  async execute(runId: string, signal?: AbortSignal): Promise<void> {
    const state = this.store.loadState(runId);
    if (!state) return;
    const requestData = state.request as Record<string, any>;
    const req = restoreRequest(requestData, SECRET_VAULT.get(runId) || {});
    const suite = state.suite as Record<string, any>;
    const instances = byDifficulty(suite.instances as Array<Record<string, any>>);

    const pconf = req.provider!;
    const role = pconf.role || 'provider';

    try {
      throwIfAborted(signal);
      this.store.updateState(runId, { status: 'running', phase: '评测中' });

      for (const inst of instances) {
        await sleep(0);
        throwIfAborted(signal);
        await this.runOnce(runId, inst, pconf, role, 0, 'main', req.scaffold, req.turn_limit, signal);
        const st = this.store.loadState(runId)!;
        st.counts.runs = (st.counts.runs ?? 0) + 1;
        st.progress.done = st.counts.runs;
        st.progress.current = null;
        this.store.updateState(runId, { counts: st.counts, progress: st.progress });
      }

      this.store.updateState(runId, { status: 'analyzing', phase: '分析汇总' });
      const records = this.store.loadRecords(runId);
      const report = analyzer.buildReport(runId, records, suite, req);
      const csvText = recordsToCsv(records);
      const mdText = analyzer.renderMarkdown(report);
      this.store.writeReport(runId, report, csvText, mdText);
      const st = this.store.loadState(runId)!;
      st.progress.current = null;
      this.store.updateState(runId, {
        status: 'completed',
        phase: '完成',
        report_ready: true,
        progress: st.progress,
        error: null,
        ended_at: new Date().toISOString().replace(/\.\d{3}Z$/, 'Z'),
      });
    } catch (err) {
      if (signal?.aborted || (err as Error).name === 'AbortError' || (err as any)?.__cancelled) {
        this.store.updateState(runId, {
          status: 'cancelled',
          phase: '已取消',
          ended_at: new Date().toISOString().replace(/\.\d{3}Z$/, 'Z'),
        });
        throw err;
      }
      const message = err instanceof Error ? (err.stack || err.message).slice(-2000) : String(err);
      this.store.updateState(runId, {
        status: 'failed',
        phase: '失败',
        error: message,
        ended_at: new Date().toISOString().replace(/\.\d{3}Z$/, 'Z'),
      });
    }
  }

  private async runOnce(
    runId: string,
    instDict: Record<string, any>,
    pconf: ProviderConfig,
    role: string,
    runIndex: number,
    phase: 'main',
    scaffold: string,
    turnLimit: number,
    signal?: AbortSignal,
  ): Promise<Record<string, any>> {
    const inst = toInstance(instDict);
    const { system, user } = buildPrompt(inst);
    const started = new Date().toISOString();

    const wsOptions: AgentWorkspaceOptions = {
      docker: true,
      image: inst.docker_image,
      repoDir: inst.repo_directory || undefined,
    };

    const st = this.store.loadState(runId)!;
    st.progress.current = { instance_id: inst.instance_id, role, phase };
    const istat = (st.instance_status[inst.instance_id] ||= {});
    istat[`${role}_r${runIndex}`] = 'running';
    this.store.updateState(runId, { progress: st.progress, instance_status: st.instance_status });

    const liveKey = livemod.register(
      runId,
      inst.instance_id,
      role,
      runIndex,
      pconf.model,
      phase,
      {
        price_input_per_m: pconf.price_input_per_m,
        price_cached_per_m: pconf.price_cached_per_m,
        price_output_per_m: pconf.price_output_per_m,
      },
    );
    const agentMeta: Record<string, any> = {};

    const [completion, meta] = await this.agentLoop(
      runId,
      inst,
      pconf,
      role,
      runIndex,
      liveKey,
      turnLimit,
      signal,
      wsOptions,
    );
    Object.assign(agentMeta, meta);
    // 确保最终 usage/速率/花费进入实时缓冲(流式未触发或最后一段被节流时兜底)
    livemod.updateUsage(
      liveKey,
      {
        prompt_tokens: completion.prompt_tokens,
        completion_tokens: completion.completion_tokens,
        cached_tokens: completion.cached_tokens,
      },
      {
        decode_tps: completion.decode_tps,
        elapsed_s: completion.wall_s,
      },
    );
    livemod.finish(liveKey, completion.finish_reason, completion.errors);
    const outcome = await evaluateOfficialDocker(inst, completion.text, {
      image: inst.docker_image,
      repoDir: inst.repo_directory || undefined,
    });

    const cost = computeCost(
      pconf,
      completion.prompt_tokens,
      completion.cached_tokens,
      completion.completion_tokens,
    );
    const record: Record<string, any> = {
      run_index: runIndex,
      phase,
      scaffold,
      instance_id: inst.instance_id,
      provider_role: role,
      provider_name: pconf.name,
      model: pconf.model,
      status: completion.ok ? 'ok' : 'error',
      resolved: completion.ok ? Boolean(outcome.resolved) : null,
      f2p_passed: outcome.f2p_passed,
      f2p_total: outcome.f2p_total,
      p2p_passed: outcome.p2p_passed,
      p2p_total: outcome.p2p_total,
      usage: {
        prompt_tokens: completion.prompt_tokens,
        completion_tokens: completion.completion_tokens,
        cached_tokens: completion.cached_tokens,
      },
      ttft_s: completion.ttft_s,
      wall_s: Math.round(completion.wall_s * 1000) / 1000,
      decode_tps: completion.decode_tps,
      finish_reason: completion.finish_reason,
      truncated: completion.finish_reason === 'length',
      turns_used: agentMeta.turns ?? 0,
      tool_calls: agentMeta.tool_calls ?? 0,
      tool_errors:
        (completion.ok ? 0 : completion.errors.length) + (agentMeta.tool_errors ?? 0),
      errors: completion.errors,
      cost_usd: Math.round(cost * 1000000) / 1000000,
      eval_method: outcome.method,
      patch_excerpt: (completion.text || '').slice(0, 600),
      started_at: started,
      ended_at: new Date().toISOString(),
    };
    this.store.appendRecord(runId, record);

    const st2 = this.store.loadState(runId)!;
    const istat2 = (st2.instance_status[inst.instance_id] ||= {});
    const badge = completion.ok ? (outcome.resolved ? 'pass' : 'fail') : 'error';
    istat2[`${role}_r${runIndex}`] = badge;
    istat2[role] = runIndex === 0 ? badge : istat2[role] ?? badge;
    this.store.updateState(runId, { instance_status: st2.instance_status });

    if (badge === 'fail' || badge === 'error') {
      this.store.saveTrajectory(runId, inst.instance_id, role, runIndex, {
        instance_id: inst.instance_id,
        provider_role: role,
        model: pconf.model,
        run_index: runIndex,
        phase,
        scaffold,
        eval_detail: outcome.detail,
        prompt_system: system,
        prompt_user: user,
        response: completion.text,
        reasoning: completion.reasoning,
        agent: {
          turns: agentMeta.turns,
          tool_calls: agentMeta.tool_calls,
          tool_errors: agentMeta.tool_errors,
          submitted: agentMeta.submitted,
        },
        trajectory: agentMeta.transcript,
        record,
      });
    }
    return record;
  }

  private async agentLoop(
    runId: string,
    inst: Instance,
    pconf: ProviderConfig,
    role: string,
    runIndex: number,
    liveKey: string,
    turnLimit: number,
    signal?: AbortSignal,
    wsOptions: AgentWorkspaceOptions = {},
  ): Promise<[CompletionResult, Record<string, any>]> {
    const safeId = inst.instance_id.replace(/[/:]/g, '_');
    const ws = new AgentWorkspace(
      path.join(this.store.runDir(runId), 'workspaces', safeId, `${role}_r${runIndex}`),
      wsOptions,
    );
    // seed 内含可能的镜像拉取(耗时可达数分钟)，先在进度里提示前端
    const seedState = this.store.loadState(runId)!;
    const seedCurrent = seedState.progress.current;
    if (seedCurrent && seedCurrent.instance_id === inst.instance_id) {
      seedCurrent.phase = 'pull_image';
      this.store.updateState(runId, {
        phase: '正在拉取镜像',
        progress: seedState.progress,
      });
    }
    ws.seed(inst);
    if (seedCurrent && seedCurrent.instance_id === inst.instance_id) {
      seedCurrent.phase = 'main';
      this.store.updateState(runId, {
        phase: '评测中',
        progress: seedState.progress,
      });
    }
    const provider = new LiveProvider(pconf);
    const system = AGENT_SYSTEM_TMPL.replace('{turns}', String(turnLimit));
    const { user } = buildPrompt(inst);
    const fullUser =
      user +
      '\n\nThe full task specification is also available in TASK.md inside the workspace.';
    const messages: Array<Record<string, unknown>> = [
      { role: 'system', content: system },
      { role: 'user', content: fullUser },
    ];

    const completion = createCompletionResult();
    let promptTokens = 0;
    let completionTokens = 0;
    let cachedTokens = 0;
    let toolCallCount = 0;
    let toolErrors = 0;
    let turnsUsed = 0;
    let submitted = false;
    let truncated = false;
    let lastText = '';
    const reasoningParts: string[] = [];
    const transcript: Array<Record<string, unknown>> = [];
    const started = performance.now();

    for (let turn = 1; turn <= turnLimit; turn++) {
      await sleep(0);
      throwIfAborted(signal);
      livemod.beginTurn(liveKey, `Turn ${turn}/${turnLimit}`);
      const stepBasePrompt = promptTokens;
      const stepBaseCompletion = completionTokens;
      const stepBaseCached = cachedTokens;
      const step = await provider.agentStep(messages, AGENT_TOOLS, {
        onDelta: (kind, piece) => livemod.appendText(liveKey, kind, piece),
        onStats: (u) =>
          livemod.updateUsage(
            liveKey,
            {
              prompt_tokens: stepBasePrompt + u.prompt_tokens,
              completion_tokens: stepBaseCompletion + u.completion_tokens,
              cached_tokens: stepBaseCached + u.cached_tokens,
            },
            {
              decode_tps: u.decode_tps,
              elapsed_s: (performance.now() - started) / 1000,
            },
          ),
        signal,
      });
      promptTokens += step.prompt_tokens;
      completionTokens += step.completion_tokens;
      cachedTokens += step.cached_tokens;
      livemod.updateUsage(
        liveKey,
        {
          prompt_tokens: promptTokens,
          completion_tokens: completionTokens,
          cached_tokens: cachedTokens,
        },
        {
          decode_tps: step.decode_tps,
          elapsed_s: (performance.now() - started) / 1000,
        },
      );
      if (step.ttft_s != null && completion.ttft_s == null) completion.ttft_s = step.ttft_s;
      if (!step.ok) {
        completion.ok = false;
        completion.errors.push(...step.errors);
        transcript.push({ role: 'assistant', error: step.errors });
        break;
      }
      turnsUsed = turn;
      if (step.finish_reason === 'length') truncated = true;
      if (step.text) lastText = step.text;
      if (step.reasoning) reasoningParts.push(`〔Turn ${turn}〕\n${step.reasoning.slice(0, 4000)}`);
      messages.push(step.assistant_message!);
      transcript.push(step.assistant_message!);

      if (step.tool_calls.length > 0) {
        for (const call of step.tool_calls) {
          toolCallCount += 1;
          const brief = call.arguments.slice(0, 120).replace(/\n/g, ' ');
          // 工具执行可能长达数十秒，先呈现"执行中"，结果就绪后再追加输出
          livemod.appendText(liveKey, 'tool', `⚙ ${call.name}(${brief}) ⏳ 执行中…\n`);
          const [resultText, isErr] = await ws.executeTool(call.name, call.arguments, signal);
          livemod.appendText(
            liveKey,
            'tool',
            `${isErr ? '✗' : '✓'} ${resultText.slice(0, 600)}\n\n`,
          );
          if (isErr) toolErrors += 1;
          if (signal?.aborted) break;
          const toolMsg: Record<string, unknown> = {
            role: 'tool',
            tool_call_id: call.id,
            content: resultText,
          };
          messages.push(toolMsg);
          transcript.push({
            role: 'tool',
            tool_call_id: call.id,
            name: call.name,
            content: resultText.slice(0, 2000),
          });
          if (call.name === 'submit') {
            submitted = true;
            break;
          }
        }
        if (signal?.aborted || submitted) break;
      } else {
        const note =
          'You did not call any tool. Use the tools to work on the task, or call `submit` when you are done.';
        messages.push({ role: 'user', content: note });
        livemod.appendText(liveKey, 'content', '(未调用工具 — 已追加行动提醒)\n');
      }
    }
    if (signal?.aborted) {
      completion.ok = false;
      completion.errors.push('cancelled');
    }


    let patch = ws.finalPatch();
    ws.dispose();
    if (!patch.trim() && lastText.trim()) patch = lastText;
    completion.text = patch;
    completion.reasoning = reasoningParts.join('\n\n');
    completion.prompt_tokens = promptTokens;
    completion.completion_tokens = completionTokens;
    completion.cached_tokens = cachedTokens;
    completion.wall_s = (performance.now() - started) / 1000;
    completion.finish_reason = truncated && !patch.trim() ? 'length' : 'stop';
    completion.decode_tps =
      completionTokens > 0 && completion.ttft_s != null && completion.wall_s > completion.ttft_s
        ? completionTokens / (completion.wall_s - completion.ttft_s)
        : null;

    const meta: Record<string, any> = {
      turns: turnsUsed,
      tool_calls: toolCallCount,
      tool_errors: toolErrors,
      submitted,
      transcript,
    };
    return [completion, meta];
  }
}

function restoreRequest(data: Record<string, any>, secrets: Record<string, string>): RunRequest {
  function fix(p: Record<string, any> | null | undefined): ProviderConfig | null {
    if (!p) return null;
    const copy = { ...p };
    const key = secrets.provider;
    if (key) copy.api_key = key;
    copy.role = 'provider';
    return copy as ProviderConfig;
  }
  return {
    provider: fix(data.provider),
    suite_level: data.suite_level ?? 'smoke6',
    suite_seed: data.suite_seed ?? 0,
    suite_id: data.suite_id ?? null,
    scaffold: 'agent',
    turn_limit: data.turn_limit ?? 50,
    dataset_source: data.dataset_source ?? 'official',
    docker_enabled: true,
    evaluator: 'official',
  };
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function throwIfAborted(signal?: AbortSignal): void {
  if (signal?.aborted) {
    const err = new Error('run cancelled') as Error & { __cancelled?: boolean };
    (err as any).__cancelled = true;
    throw err;
  }
}
